// @optimizely-opal/opal-tools-sdk/dist/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Function = exports.AuthRequirement = exports.Parameter = exports.ParameterType = void 0;
/**
 * Types of parameters supported by Opal tools
 */
var ParameterType;
(function (ParameterType) {
    ParameterType["String"] = "string";
    ParameterType["Integer"] = "integer";
    ParameterType["Number"] = "number";
    ParameterType["Boolean"] = "boolean";
    ParameterType["List"] = "list";
    ParameterType["Dictionary"] = "object";
})(ParameterType || (exports.ParameterType = ParameterType = {}));
/**
 * Parameter definition for an Opal tool
 */
class Parameter {
    /**
     * Create a new parameter definition
     * @param name Parameter name
     * @param type Parameter type
     * @param description Parameter description
     * @param required Whether the parameter is required
     */
    constructor(name, type, description, required) {
        this.name = name;
        this.type = type;
        this.description = description;
        this.required = required;
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        return {
            name: this.name,
            type: this.type,
            description: this.description,
            required: this.required
        };
    }
}
exports.Parameter = Parameter;
/**
 * Authentication requirements for an Opal tool
 */
class AuthRequirement {
    /**
     * Create a new authentication requirement
     * @param provider Auth provider (e.g., "google", "microsoft")
     * @param scopeBundle Scope bundle (e.g., "calendar", "drive")
     * @param required Whether authentication is required
     */
    constructor(provider, scopeBundle, required = true) {
        this.provider = provider;
        this.scopeBundle = scopeBundle;
        this.required = required;
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        return {
            provider: this.provider,
            scope_bundle: this.scopeBundle,
            required: this.required
        };
    }
}
exports.AuthRequirement = AuthRequirement;
/**
 * Function definition for an Opal tool
 */
class Function {
    /**
     * Create a new function definition
     * @param name Function name
     * @param description Function description
     * @param parameters Function parameters
     * @param endpoint API endpoint
     * @param authRequirements Authentication requirements (optional)
     */
    constructor(name, description, parameters, endpoint, authRequirements) {
        this.name = name;
        this.description = description;
        this.parameters = parameters;
        this.endpoint = endpoint;
        this.authRequirements = authRequirements;
        /**
         * HTTP method for the endpoint (default: POST)
         */
        this.httpMethod = 'POST';
    }
    /**
     * Convert to JSON for the discovery endpoint
     */
    toJSON() {
        const result = {
            name: this.name,
            description: this.description,
            parameters: this.parameters.map(p => p.toJSON()),
            endpoint: this.endpoint,
            http_method: this.httpMethod
        };
        if (this.authRequirements && this.authRequirements.length > 0) {
            result.auth_requirements = this.authRequirements.map(auth => auth.toJSON());
        }
        return result;
    }
}
exports.Function = Function;
